### PR TITLE
dhall-to-{json,yaml,yaml-ng}: Handle empty maps correctly

### LINKED ISSUE #1559: Compile empty association list to object rather than list?
Right now, unless `--noMaps` is specified, we expect `dhall-to-json` and `dhall-to-yaml` to compile values of type `List { mapKey = t, mapValue = u }` to objects. This works when the list is populated, but it seems that whatever process transforms non-empty association lists to objects is skipped for empty ones, so you end up with a different type in the compiled file. For example:

    $ dhall-to-json --version
    1.5.0

    $ dhall-to-json <<< '[{ mapKey = "foo", mapValue = "bar" }]'
    {
      "foo": "bar"
    }

    $ dhall-to-json <<< '[] : List { mapKey : Text, mapValue : Text }'
    []

Expected:

    $ dhall-to-json <<< '[] : List { mapKey : Text, mapValue : Text }'
    {}

If this is actually the desired behavior, please let me know! But if it is, I think it would still make sense to add this behavior as an option (maybe add a command line flag like `--transformEmptyAssocs` or something).

There are definitely practical applications of this new behavior. In my case, I'm compiling a docker compose configuration where I represent services and networks and such as association lists, e.g.

    { services = [
      { mapKey = "my_service"
      , mapValue = Service::
        { image = "dhallhaskell/dhall-json"
        , ports = [ "8000:80" ]
        }
      } ]
    , networks = [
      { mapKey = "my_net", mapValue = Network.default } ]
    }
      
==>

    services:
      my_service:
        image: "dhallhaskell/dhall-json"
        ports: [ "8000:80" ]
    networks:
      my_net:
        driver: overlay

In my case, if no, say, networks have been specified, it needs to be an object rather than a list, or docker will reject it:

    networks: {}

I have a work-around using `Optional`, but the ergonomics are **dramatically** better if I can allow `networks = [] : List { mapKey = Text, mapValue = Network.Type }` to be empty.

### LINKED ISSUE #1560: Differing formatting for empty lists with dhall-to-yaml and dhall-to-yaml-ng
```
$ echo '[] : List Bool' | cabal run dhall-to-yaml
 []
$ echo '[] : List Bool' | cabal run dhall-to-yaml-ng
[]
```

Note the leading space with `dhall-to-yaml`.

@patrickmn Could you take a look?

### PR DESCRIPTION
**Pull request by sjakobi**
While generalizing list annotations In https://github.com/dhall-lang/dhall-haskell/commit/0ee6ce6f6fd2afe9d227fa9d58a3e61fced33c21, I had forgotten to update `dhall-json`.

Fixes #1559.

This requires aeson-yaml >= 1.0.5.0 since older versions
incorrectly encode empty objects.

The raised bound also fixes #1560. A regression test is included.

### CODE REVIEW

**Comment by sjakobi** [comment]
Sadly `dhall-to-yaml` and `dhall-to-yaml-ng` disagree about how to encode empty objects:

```
  ./tasty/data/emptyMap
    HsYAML:     OK
    aeson-yaml: FAIL
      tasty/Main.hs:65:
      Conversion to YAML did not generate the expected output
      expected: "{}\n"
       but got: "\n"
```

Could you take a look at that @patrickmn?

File: dhall-yaml/tasty/Main.hs [hunk 1]
```diff
@@ -32,6 +33,10 @@ testTree =
         , testDhallToYaml
             Dhall.JSON.Yaml.defaultOptions
             "./tasty/data/special"
+        , testDhallToYaml
+            Dhall.JSON.Yaml.defaultOptions
+                { conversion = Dhall.JSON.Conversion "mapKey" "mapValue" }
```
Comment (sjakobi) [inline dhall-yaml/tasty/Main.hs]: I wonder whether we should update these `defaultOptions` to accurately reflect the default CLI options.

(In `testDhallToJSON`, we already use this `Conversion` by default:

https://github.com/dhall-lang/dhall-haskell/blob/6c68f8206df5dce9318a23116a5bd3454b8582cd/dhall-json/tasty/Main.hs#L69-L74

)

Comment (sjakobi) [reply dhall-yaml/tasty/Main.hs]: I have prepared [a patch](https://github.com/dhall-lang/dhall-haskell/commit/3cbf18ff2cc8957d52a9565360ebb5ee49079e62) that adds a `defaultConversion`, that we could apply after this fix.

**Comment by patrickmn** [comment]
Yep, will fix.

**Comment by sjakobi** [comment]
> Yep, will fix.

Ping @patrickmn! :)

**Comment by sjakobi** [comment]
@patrickmn Any chance this could make it into the new release (https://github.com/dhall-lang/dhall-haskell/pull/1575)?

**Comment by patrickmn** [comment]
Sorry, wasn't here much lately.

Yes, if it's just the empty objects and not the single vs. double-quoting right now I can push a change and release in the next hour or so.

**Comment by sjakobi** [comment]
No worries @patrickmn. I would really appreciate a fix for the empty-object encoding! :)

#1516 is less important IMHO.

**Comment by patrickmn** [comment]
Ok, should be fixed in 1.0.5.0 - https://github.com/clovyr/aeson-yaml/commit/2b4abb13eb8cfe35154f3d321e44286f2847f97f

**Comment by sjakobi** [comment]
Thanks for the fix @patrickmn! :)

File: dhall-yaml/tasty/Main.hs (hunk 1 above)
Comment (Gabriel439) [reply dhall-yaml/tasty/Main.hs]: Yeah, it would be a good idea to keep them in sync

**Pull request merged**